src/wg/image/imagebuffer.d

```diff
@@ -24,6 +28,68 @@ struct ImageBuffer
 }
 
 
+/**
+ * Allocate an image buffer for the specified format in GC memory.
+ * Optionally allocate additional metadata pages.
+ */
+ImageBuffer allocImage(MetadataBlocks...)(const(char)[] format, uint width, uint height, size_t[] additionalMetadataBytes...)
+{
+    import std.exception : enforce;
+
+    assert(additionalMetadataBytes.length == MetadataBlocks.length);
+
+    ImageBuffer img;
+    void[] mem = allocImageImpl(format, true, width, height, getMetadataSize!(false, MetadataBlocks)(additionalMetadataBytes), getGcAllocator(), img);
+    enforce(mem.length > 0, "Failed to allocate image with format: " ~ format);
+    arrangeMetadata!(false, MetadataBlocks)(img.metadata, additionalMetadataBytes);
+    return img;
+}
+
+/**
+ * Allocate an image buffer for the specified format using the supplied allocator.
+ * Optionally allocate additional metadata pages.
+ */
+ImageBuffer allocImage(MetadataBlocks...)(const(char)[] format, uint width, uint height, Allocator* allocator, size_t[] additionalMetadataBytes...) nothrow @nogc
+{
+    assert(additionalMetadataBytes.length == MetadataBlocks.length);
+
+    ImageBuffer img;
+    void[] mem = allocImageImpl(format, true, width, height, getMetadataSize!(true, MetadataBlocks)(additionalMetadataBytes), allocator, img);
+    if (mem.length)
+    {
+        AllocationMetadata* allocData = arrangeMetadata!(true, MetadataBlocks)(img.metadata, additionalMetadataBytes);
+        allocData.allocations[0].allocator = allocator;
+        allocData.allocations[0].mem = mem;
+    }
+    return img;
+}
+
+/**
+ * Free any allocations contained in an image buffer.
+ */
+void freeImage(ref ImageBuffer image) nothrow @nogc
+{
+    AllocationMetadata* allocData = image.getMetadata!AllocationMetadata();
+    if (allocData)
+    {
+        // cache the allocation pages locally, since we will probably delete them along the way
+        AllocationMetadata.Page[128] allocations;
```

Comment:
Probably nitpicking but I would add check that `allocData.allocations.length` does not exceed `allocations.length`

Reply:
oh yeah, good point ;)